src/jdk.incubator.code/share/classes/jdk/incubator/code/internal/CodeModelToAST.java

```diff
@@ -0,0 +1,287 @@
+        var paramTypes = invokeOp.invokeDescriptor().type().parameterTypes().stream().map(this::typeElementToType)
+                .collect(List.collector());
+        var restype = typeElementToType(invokeOp.resultType());
+        var type = new Type.MethodType(paramTypes, restype, List.nil(), syms.methodClass);
+        var methodSym = new Symbol.MethodSymbol(flags, name, type,
+                typeElementToType(invokeOp.invokeDescriptor().refType()).tsym);
+        var meth = receiver == null ? treeMaker.Ident(methodSym) : treeMaker.Select((JCTree.JCExpression) valueToTree.get(receiver), methodSym);
+        var args = new ListBuffer<JCTree.JCExpression>();
+        for (Value operand : arguments) {
+            args.add((JCTree.JCExpression) valueToTree.get(operand));
+        }
+        var methodInvocation = treeMaker.App(meth, args.toList());
+        if (invokeOp.isVarArgs()) {
+            var lastParam = invokeOp.invokeDescriptor().type().parameterTypes().getLast();
+            Assert.check(lastParam instanceof ArrayType);
+            methodInvocation.varargsElement = typeElementToType(((ArrayType) lastParam).componentType());
+        }
+        if (invokeOp.result().uses().isEmpty()) {
+            return treeMaker.Exec(methodInvocation);
+        }
+        return methodInvocation;
+    }
+
+    private JCTree opToTree(Op op) {
+        JCTree tree = switch (op) {
+            case CoreOp.ConstantOp constantOp when constantOp.value() == null ->
+                    treeMaker.Literal(TypeTag.BOT, null).setType(syms.botType);
+            case CoreOp.ConstantOp constantOp -> treeMaker.Literal(constantOp.value());
+            case CoreOp.InvokeOp invokeOp -> invokeOpToJCMethodInvocation(invokeOp);
+            case CoreOp.NewOp newOp -> {
+                if (newOp.resultType() instanceof ArrayType at) {
+                    var elemType = treeMaker.Ident(typeElementToType(at.componentType()).tsym);
+                    // @@@ we assume one dimension
+                    var dims = List.of((JCTree.JCExpression) valueToTree.get(newOp.operands().getFirst()));
+                    var na = treeMaker.NewArray(elemType, dims, null);
+                    na.type = typeElementToType(at);
+                    yield na;
+                }
+                var ownerType = typeElementToType(newOp.constructorType().returnType());
+                var clazz = treeMaker.Ident(ownerType.tsym);
+                var args = new ListBuffer<JCTree.JCExpression>();
+                for (Value operand : newOp.operands()) {
+                    args.add((JCTree.JCExpression) valueToTree.get(operand));
+                }
+                var nc = treeMaker.NewClass(null, null, clazz, args.toList(), null);
+                var argTypes = new ListBuffer<Type>();
+                for (Value operand : newOp.operands()) {
+                    argTypes.add(typeElementToType(operand.type()));
```

Comment:
This seems wrong -- and is related to the point I made above -- the type of the operands here are the types of the actual arguments of the constructor calls. So, if you have:
```
class Foo {
     Foo(String string) { ... }
 }
 ```
 
 the method type of the constructor call has to say `Foo(String)`. But if you "infer" the type of the constructor argument from the actual argument, if the call is like this:
 
 ```
 new Foo(null)
 ```
 
We should probably use the op's `constructorType` to derive the type of the invoked constructor? (You use that to determine the constructor's owner type)